debian/naivecalendar.1
.TH NAIVECALENDAR 1 "January 6, 2021"

.SH NAME
naivecalendar 
- rofi calendar popup

.SH SYNOPSIS
.B naivecalendar [-h] [-V] [-v] [-p] [-x] [-f 
.I FORMAT
.B ] [-e 
.I EDITOR
.B ] [-l 
.I LOCALE
.B ] [-c] [-t 
.I THEME
.B ] [-d
.I DATE
.B ]

.SH DESCRIPTION
.B naivecalendar
is made using rofi and python3 stdlib.
it can be used as an interactive calendar or a simple popup to pickup date

.RS
.nf
 - Use mouse or keyboard to interact with the calendar.
 - Hit bottom arrows to cycle through months.
 - Hit a day to create a linked event.
(A day with attached event will appear yellow.)
 - Create multiple event type and with between them

There's somme shortcut too, type it in rofi prompt :

     help : display this help
 nn or -- : go to previous year
   n or - : go to previous month
   p or + : go to next month
 pp or ++ : go to next year
    event : display events of the month (first line)
   switch : switch events folder to display
    theme : show theme selector
     menu : display a selection menu (skip shortcuts)

.SH OPTIONS

.TP
.B -h|--help 
show this help message and exit

.TP
.B -V|--version 
show program's version number and exit

.TP
.B -v|--verbose 
redirect rofi command errors to stdout

.TP
.B -p|--print
print date to stdout instead of opening a note

.TP
.B -x|--clipboard
copy date to clipboard

.TP
.BI -f|--format " "FORMAT
option '-p' or '-x' output format (datetime.strftime format, defaut='%Y-%m-%d')

.TP
.BI -e|--editor " "EDITOR
editor command to open notes

.TP
.BI -l|--locale " "LOCALE
force system locale, for example '-l es_ES.utf8'

.TP
.B -c|--read-cache      
force calendar to read old date from cache

.TP
.BI -t|--theme " "THEME
set calendar theme, default=classic_dark (theme file name without extention)

.TP
.BI -d|--date " "DATE  
display calendar at the given month, format='%m-%Y'



.SH FILES
.SS System commands and scripts files

.TP
.I /usr/bin/naivecalendar
.RS
system command                     

.RE
.I /usr/share/naivecalendar/naivecalendar.sh
.RS
rofi command                       

.RE
.I /usr/share/naivecalendar/naivecalendar.py
.RS
script called by rofi              

.RE
.I /usr/share/naivecalendar/tools/naivecalendar-update-themes
.RS
theme config editor system command        

.RE
.I /usr/share/naivecalendar/tools/naivecalendar-add-event
.RS
theme event editor system command         

.RE
.SS Configuration Files

.TP
.I /usr/share/naivecalendar/themes/*.rasi
.RS
rofi configuration theme files                   

.RE
.I /usr/share/naivecalendar/themes/common/
.RS
common ressources for themes (rofi configuration files)

.RE
.I /usr/share/naivecalendar/themes/*.cfg
.RS
ini file for calendar content configuration     

.RE
.I /usr/share/naivecalendar/global/events.cfg   
.RS
ini file to configure events                        

.RE
.I /usr/share/naivecalendar/global/custom_actions.cfg
.RS
ini file to configure custom buttons and actions                     

.RE
.I /usr/share/naivecalendar/scripts/"    
.RS
default folder to store custom actions scripts

.RE

.SS User configurations files

.TP
.I ~/.config/naivecalendar/themes/*.rasi
.RS
user theme files                   

.RE
.I ~/.config/naivecalendar/themes/*.cfg
.RS
user content configuration         

.RE
.I ~/.config/naivecalendar/themes/common/
.RS
common ressources for user themes

.RE
.I ~/.config/naivecalendar/global/custom_actions.cfg
.RS
user custom actions                

.RE
.I ~/.config/naivecalendar/global/events.cfg   
.RS
user events configuration

.RE
.I ~/.config/naivecalendar/scripts/"    
.RS
user scripts                       

.RE

.SS Generated and cache files

.TP
.I ~/.naivecalendar_events/<event type>/<date format>.txt
.RS
day notes path (default)           

.RE
.I ~/.cache/naivecalendar/date_cache.ini
.RS
remember date throught loops       

.RE
.I ~/.cache/naivecalendar/pretty_print_cache.txt
.RS
pass date to bash when -p option   

.RE
.I ~/.cache/naivecalendar/theme_cache.txt
.RS
remember theme after quitting      

.RE
.I ~/.cache/naivecalendar/event_cache.txt
.RS
remember event type                

.RE
.I ~/.cache/naivecalendar/rofi_log.txt
.RS
last rofi command log file         


Changelog
v2.0.1
	Major:
		Update themes file for rofi 1.7 compatibility
	Minor:
		Add themes that follow user's ones
		New update option : clone => populate user config folder
	Other:
		Add Makefile 

v2.0.0
	Major:
		/!\ Break conig files compatibility (rename variable,
		see xxx to upgrade your files
		Add Custom event support
		Add CalDav support (as custom event)
		Simplified theme conf : remove hardly ever used variable, now hardcoded
		Move event config in separated file and common to all themes

	Minor:
	    Improve conf files doc
	    New color theme : material red

v1.0.0

	Major:
		Release naivecalendar as first stable

v0.8.1

	Minor:
		Workaround for "back to current date"

v0.8.0 :

	Major:
		Allow user to use full pango markup (+1versatility, +1kiss)
		(~soft breaking)
	Minor:
		Allow custom character in ROWs
		Add themes : pastel blue, slate
		theme ressources haromonization (standard?)
		remove useless [SHAPE] section in config (now just count ROWs)
		Doc adjustements
		Change default theme to classic dark extended (better example)

v0.7.4 : Opulence

	Major:
		Theme is also applied to popups and menus
	Minor:
		Change datetime : today() to now() : can display hour in header/message
		Add Mangolia and Carbon Gold themes from ZB652
		Add shortcut => back to current month
		Correct code blocks in doc
		Update completion

v0.7.3 : Quinquennial plan

	Minor:
		Add "modern" themes
		Revert log file (correct recursion effect)
	Bugs:
		missing return in set_list function
		typing error when searching for theme existence in user path

v0.7.2 : Socialism

	Minor:
		Themes now share common ressources
	Bugs:
		log was erased at last loop
		reopen notes with "strftime folder" in show event popup
		
v0.7.1 : Faith

	Minor:
		Capture rofi close error when openning menus
		Save session cmd parameters after openning menu
		Events now support strftime format in folder name

	Bugs:
		Only execute -p and -x action once (not in recursions)		
		Unwanted blank line in output

v0.7.0 : Again and again and again

	Major:
		Add config option : update message at each date change
		Event : tiny format to parse event for menus
		
	Minor:
		New themes : lead pencil and blackboard
		Update readme : add version number
		Update doc : add example
		Keep whitespace defined in config for alignement control
		Can now set "back to calendar" symbol

v0.6.2 : Improvements

    Minor:
        Improve Completion
        Update Readmes

    Bugs:
        infinite loop with --theme option
        special symbols issues in regex

v0.6.1 : Scripts & Completion

    Major:
        add completion
    Minor:
        improve scripts
    Bugs:
        script now handle special character as space and slash

v0.6.0 : Recursion

    Major:
        Remove unconsistent mode-switcher
        Add script to update all theme at once
        Rewrite menu and shortcut bar construction (prevent unconsitent config)
    Minor:
        Add missing optional dependencies
        Prevent recursive accumulation of rofi calls
    Bugs:
        Prevent crash when wrong theme name is given

v0.5.2 : Themes Conflicts

    Major:
        Now override all theme.rasi files, naivecalendat should appear the same on all configs

v0.5.1 : Help & Readme

    Minor:
        Better help msg
        Update Readme

v0.5.0 : Better header and themes

    Major :
        More configurable header (break old config)
    Minor :
        New themes

V0.4.1 : Alignment

    Minor:
        Align all symbols/character to right

v0.4.0 : Menu & Shortcuts

    Major:
        Add a new bar that can be used for shortcuts and menu
        Reformat cfg files : breaking changes
    Minor:
        Update and add themes variant
        Document config files
    Bugs:
        augment sleep time before relaunch rofi to prevent "lose focus click"
        prevent crash when conflict between old cache file and new config

v0.3.5 : Multiple Events

    Major:
        Add multiple events folder (replace notes)
        Add a menu button
        Better symbols support for event files names
    Minor:
        Update doc / man / readme
    Bugs:
        support name with point for notes (now events)

v0.3.4 : Hello Copy-to-clipboard

    Major:
        add copy-to-clipboard (-x|--clipboard) option
        add version message (-v|--version)
    Minor:
        docstring, manual and message update
        more explicit global var assignement

v0.3.3 : Byebye tkinter

    Major:
        remove tkinter dependancies (rofi popup only)
    Minor:
        document undocumented functions

v0.3.2 : The unamed version

    Major :
        add "open at date" option
    Minor :
        correct numpy docstrings
    Bugs :
        missing dep in deb
        non-recursive config folder creation⎋

v0.3.1 : Debpackage

    Major:
        Create debian package


src/naivecalendar.sh
#!/bin/bash

param="$(printf ' %q' "$@")"
cmd="${BASH_SOURCE%/*}/naivecalendar.py $(printf ' %q' "$@")"

##############################################################################
####################  Process command line arguments #########################
##############################################################################


while [ $# -gt 0 ]; do
    key="$1"
    case $key in
      -h|--help)
        $cmd
        exit 0
        ;;
      -V|--version)
        $cmd
        exit 0
        ;;
      -t|--theme)
        OPT_THEME="$2"
        shift
        shift
        ;;
      -v|--verbose)
        IS_VERBOSE='true'
        shift
        ;;
      -x|--clipboard)
        IS_CLIPBOARD='true'
        shift
        ;;
      -p|--print)
        IS_PRINT='true'
        shift
        ;;
      -f|--format)
        shift
        shift
        ;;
      -e|--editor)
        shift
        shift
        ;;
      -l|--locale)
        shift
        shift
        ;;
      -d|--date)
        shift
        shift
        ;;
      -c|--read-cache)
        shift
        ;;
      *)
        echo "Argument non défini : '$key'" 
        shift
        ;;
    esac
done


# load theme file
THEME_CACHE_FILE="$HOME/.cache/naivecalendar/theme_cache.txt"
THEME_USER_PATH="$HOME/.config/naivecalendar/themes"
THEME_SOURCE_PATH="${BASH_SOURCE%/*}/themes"



# forced with cmd line argument
if ! [ -z "$OPT_THEME" ]; then
    # search in user path before
    THEME="$THEME_USER_PATH/$OPT_THEME.rasi"
    if ! test -f "$THEME"; then
        THEME="$THEME_SOURCE_PATH/$OPT_THEME.rasi"
    fi
    if ! test -f "$THEME"; then
        echo "theme '$OPT_THEME' doesn't exist"
        exit 0
    fi

# otherwise, look in cache
elif test -f "$THEME_CACHE_FILE"; then
    # look in user path, then in source path
    THEME="$THEME_USER_PATH/$(cat $THEME_CACHE_FILE).rasi"
    if ! test -f "$THEME"; then
        THEME="$THEME_SOURCE_PATH/$(cat $THEME_CACHE_FILE).rasi"
    fi
    if ! test -f "$THEME"; then
        echo "*************************************************************************"
        echo "* '$(cat $THEME_CACHE_FILE)' theme doesn't exist anymore, "
        echo "*    - please remove $THEME_CACHE_FILE    "
        echo "* or                                                                    *"
        echo "*    - force a theme with -t|--theme argument,                          *"
        echo "*      then switch again to an existing theme to override cache file    *"
        echo "*************************************************************************"
        exit 0
    fi

else
    THEME="$THEME_SOURCE_PATH/classic_dark_extended.rasi"
fi

##############################################################################
############################  call rofi  #####################################
##############################################################################

ROFI_LOG_FILE="$HOME/.cache/naivecalendar/rofi_log.txt"
mkdir -p "$(dirname $ROFI_LOG_FILE)"
touch $ROFI_LOG_FILE

# Determine if is it a "first" call or a recursion
if [ -z $NAIVECALENDAR_FIRST_CALL ]; then
    export NAIVECALENDAR_FIRST_CALL="true"
fi
if [ -z $NAIVECALENDAR_SECOND_CALL ]; then
    export NAIVECALENDAR_SECOND_CALL="true"
fi

loop_nb=-1

if [[ "$NAIVECALENDAR_FIRST_CALL" == "true" && "$NAIVECALENDAR_SECOND_CALL" == "true" ]]; then
    loop_nb=0
    export NAIVECALENDAR_FIRST_CALL="false"

    # empty log file at first call
    echo "--- end naivecalendar rofi log ---" > "$ROFI_LOG_FILE"

elif [[ "$NAIVECALENDAR_FIRST_CALL" == "false" && "$NAIVECALENDAR_SECOND_CALL" == "true" ]]; then
    loop_nb=1
    export NAIVECALENDAR_FIRST_CALL="false"
    export NAIVECALENDAR_SECOND_CALL="false"


elif [[ "$NAIVECALENDAR_FIRST_CALL" == "false" && "$NAIVECALENDAR_SECOND_CALL" == "false" ]]; then
    loop_nb=2
fi


# launch rofi
rofi_output="$(rofi -show calendar \
    -modi "calendar:$cmd" \
    -theme-str '@theme "'$THEME'"' \
    2>&1)"

# write to log file (prepend because of recursion)
printf '%s\n' "$rofi_output" | tac | while read -r line; do sed -i '1s?^?'"$line"'\n?' "$ROFI_LOG_FILE"; done

if [ $loop_nb -eq 0 ]; then
    sed -i  "1i--- first call ---" "$ROFI_LOG_FILE"
elif [ $loop_nb -eq 1 ]; then
    sed -i "1i--- first recursion ---" "$ROFI_LOG_FILE"
elif [ $loop_nb -eq 2 ]; then
    sed -i "1i--- 2+ recursion ---" "$ROFI_LOG_FILE"
fi


##############################################################################
####################  Naivecalendar ouputs ###################################
##############################################################################

# execute copy-to-clipboard or print action only with main/first script call
if [ "$loop_nb" -eq 0 ]; then

    # print log
    if [ "$IS_VERBOSE" = "true" ]; then
        cat "$ROFI_LOG_FILE"
    fi

    # print date to stdout
    if [ "$IS_CLIPBOARD" = "true" ]; then
        FILE="$HOME/.cache/naivecalendar/pretty_print_cache.txt"
        cat $FILE | xclip -selection clipboard &
    fi

    # print date to stdout
    if [ "$IS_PRINT" = "true" ]; then
        FILE="$HOME/.cache/naivecalendar/pretty_print_cache.txt"
        cat $FILE > "$HOME/temp.txt"
        printf "%s" "$(cat $FILE)"
    fi
fi

